builtins/web/event/event-target.cpp

```diff
@@ -0,0 +1,557 @@
+    return false;
+  }
+
+  // 9. If found is false and event's isTrusted attribute is true:
+  // N/A
+  return true;
+}
+
+
+// https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke
+bool EventTarget::inner_invoke(JSContext *cx, HandleObject event, JS::HandleVector<EventListener> list, bool *found) {
+  // 1. Let found be false.
+  *found = false;
+
+  // 2. For each listener of listeners, whose removed is false:
+  for (auto &listener : list) {
+    if (listener.removed) {
+      continue;
+    }
+
+    // 1. If event's type attribute value is not listener's type, then continue.
+    RootedString type_str(cx, Event::type(event));
+    auto event_type = core::encode(cx, type_str);
+    if (!event_type) {
+      return false;
+    }
+
+    if (listener.type != std::string_view(event_type)) {
+      continue;
+    }
+
+    // 2. Set found to true.
+    *found = true;
+
+    // 3. If phase is "capturing" and listener's capture is false, then continue.
+    // 4. If phase is "bubbling" and listener's capture is true, then continue.
+    // N/A
+
+    // 5. If listener's once is true, then remove an event listener given event's
+    //  currentTarget attribute value and listener.
+    if (listener.once) {
+      auto current_target = Event::current_target(event);
+      MOZ_ASSERT(is_instance(current_target));
+
+      auto target_list = listeners(current_target);
+      MOZ_ASSERT(target_list);
+
+      auto it = std::find_if(target_list->begin(), target_list->end(), [&](const auto &other) {
+        return listener == other;
+      });
+
+      if (it != target_list->end()) {
+        it->removed = true;
+        target_list->erase(it);
+      }
+    }
+
+    // 6. Let global be listener callback's associated realm's global object.
+    // 7. Let currentEvent be undefined.
+    // 8. If global is a Window object:
+    //  1. Set currentEvent to global's current event.
+    //  2. If invocationTargetInShadowTree is false, then set global's current event to event.
+    //  N/A
+    // 9. If listener's passive is true, then set event's in passive listener flag.
+    if (listener.passive) {
+      Event::set_passive_listener(event, true);
+    }
+
+    // 10. If global is a Window object, then record timing info for event listener given event and listener.
+    // N/A
+    //
+
+    // 11. Call a user object's operation with listener's callback, "handleEvent", event,
+    // and event's currentTarget attribute value.
+    auto engine = api::Engine::get(cx);
+    RootedValue callback_val(cx, listener.callback);
+    RootedObject callback_obj(cx, &callback_val.toObject());
+
+    RootedValue rval(cx);
+    RootedValueArray<1> args(cx);
+    args[0].setObject(*event);
+
+    if (JS::IsCallable(callback_obj)) {
+      auto global = engine->global();
+      if (!JS::Call(cx, global, callback_val, args, &rval)) {
+        auto msg = "dispatching event " + listener.type;
+        engine->dump_pending_exception(msg.c_str());
```

Comment:
As mentioned above, this should all be factored out to happen for all problems in the loop, and shouldn't abort the loop itself.

Also, can you file an issue for adding support for the [report an exception](https://html.spec.whatwg.org/multipage/webappapis.html#report-an-exception) concept, and add a TODO linking to it here?